File: java/terasort/1_base/src/main/java/terasort/files/Sort.java
/*
 *  Copyright 2002-2016 Barcelona Supercomputing Center (www.bsc.es)
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package terasort.files;

import java.io.File;
import java.util.LinkedList;

public class Sort {

    private static int NUM_FRAGMENTS;
    private static int NUM_RANGES;
    private static long minKey;
    private static long maxKey;
    private static String DATA_FOLDER;
    private static String OUTPUT_FOLDER;

    private static String[] filePaths;
    //private static Fragment[] result;

    public static void main(String args[]) {
        // Get parameters
        if (args.length != 2) {
            System.out.println("[ERROR] Usage: Sort <DATA_FOLDER> <OUTPUT_FOLDER>");
            System.exit(-1);
        }
        DATA_FOLDER = args[0];
        System.out.println("[LOG] DATA_FOLDER   = " + DATA_FOLDER);
        OUTPUT_FOLDER = args[1];
        System.out.println("[LOG] OUTPUT_FOLDER = " + OUTPUT_FOLDER);

        // Run sort by key app
        long startTime = System.currentTimeMillis();
        Sort sbk = new Sort();
        sbk.run();

        System.out.println("[LOG] Main program finished.");

        // Uncomment the following lines to see the result keys
        //for (int i=0; i < NUM_RANGES; ++i){
        //	System.out.println("Range " + i + " result:");
        //	System.out.println(result[i]);
        //}
        long endTime = System.currentTimeMillis();

        System.out.println("[TIMER] Elapsed time: " + (endTime - startTime) + " ms");

    }

    /**
     * Execution method.
     */
    private void run() {
        // Initialize file Names
        System.out.println("[LOG] Initialising filenames for each matrix");
        initializeVariables();

        // Compute result
        System.out.println("[LOG] Reading fragments.");
        System.out.println("[LOG] Filtering fragments.");
        Fragment[][] ranges = new Fragment[NUM_FRAGMENTS][NUM_RANGES];
        for (int i = 0; i < NUM_FRAGMENTS; ++i) {
            Fragment fragment = SortImpl.getFragment(filePaths[i]);
            long start = 0;
            long step = ((maxKey - minKey) / NUM_RANGES);
            long end = step;
            for (int j = 0; j < NUM_RANGES; ++j) {
                ranges[i][j] = SortImpl.filterTask(fragment, start, end);
                start += step;
                end += step;
            }
        }
        System.out.println("[LOG] Merging filtered results.");
        System.out.println("[LOG] Sorting fragment.");
        System.out.println("[LOG] Saving sorted fragments.");
        Integer[] sortedCount = new Integer[NUM_RANGES];
        for (int i = 0; i < NUM_RANGES; ++i) {
            // Merge-Reduce per range
            LinkedList<Integer> q = new LinkedList<Integer>();
            for (int j = 0; j < NUM_FRAGMENTS; ++j) {
                q.add(j);
            }
            int x = 0;
            while (!q.isEmpty()) {
                x = q.poll();
                int y;
                if (!q.isEmpty()) {
                    y = q.poll();
                    ranges[x][i] = SortImpl.reduceTask(ranges[x][i], ranges[y][i]);
                    q.add(x);
                }
            }
            Fragment result = SortImpl.sortPartition(ranges[x][i]);
            sortedCount[i] = SortImpl.saveFragment(result, OUTPUT_FOLDER);
        }

        // Merge-Reduce counters
        LinkedList<Integer> q = new LinkedList<Integer>();
        for (int j = 0; j < NUM_FRAGMENTS; ++j) {
            q.add(j);
        }
        int x = 0;
        while (!q.isEmpty()) {
            x = q.poll();
            int y;
            if (!q.isEmpty()) {
                y = q.poll();
                sortedCount[x] = SortImpl.reduceCount(sortedCount[x], sortedCount[y]);
                q.add(x);
            }
        }
        int count = sortedCount[x];

        System.out.println("[TERASORT] Total Sorted Elements = " + count);
    }

    /**
     * Variables initialization.
     */
    private void initializeVariables() {
        NUM_FRAGMENTS = new File(DATA_FOLDER).listFiles().length;
        filePaths = new String[NUM_FRAGMENTS];
        int i = 0;
        for (File f : new File(DATA_FOLDER).listFiles()) {
            filePaths[i] = f.getAbsolutePath();
            System.out.println("File: " + filePaths[i]);
            i = i + 1;
        }
        System.out.println("NUM_FRAGMENTS value = " + NUM_FRAGMENTS);

        //result = new Fragment[NUM_FRAGMENTS];
        NUM_RANGES = NUM_FRAGMENTS;

        byte[] mi = {0, 0, 0, 0, 0, 0, 0, 0};
        minKey = Utils.byte2long(mi);
        byte[] ma = {0, -1, -1, -1, -1, -1, -1, -1}; // 0xff = -1
        maxKey = Utils.byte2long(ma);
        System.out.println("Min key: " + minKey);
        System.out.println("Max key: " + maxKey);
    }

}

File: python/pycompss_lib/algorithms/sort_by_key/sortTotal/src/sort.py
#!/usr/bin/python
#
#  Copyright 2002-2018 Barcelona Supercomputing Center (www.bsc.es)
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

# -*- coding: utf-8 -*-

from pycompss.api.task import task


@task(returns=list)
def sort_partition(path):
    """ Sorts data, which is assumed to consists of (key, value) tuples list.
    :param path: file absolute path where the list of tuples to be sorted is located.
    :return: sorted list of tuples.
    """
    import pickle
    import operator
    f = open(path, 'r')
    data = pickle.load(f)
    f.close()
    res = sorted(list(data.items()), key=operator.itemgetter(1), reverse=False)
    return res


def sort_by_key(files_paths):
    """ Sort by key.
    :param files_paths: List of paths of the input files.
    :return: List of elements sorted.
    """
    from pycompss.api.api import compss_wait_on
    n = list(map(sort_partition, files_paths))
    res = merge_reduce(reduce_task, n)
    res = compss_wait_on(res)
    return res


@task(returns=list, priority=True)
def reduce_task(a, b):
    """ Reduce list a and list b.
        They must be already sorted.
    :param a: list.
    :param b: list.
    :return: result of merging a and b lists.
    """
    res = []
    i = 0
    j = 0
    while i < len(a) and j < len(b):
        if a[i] < b[j]:
            res.append(a[i])
            i += 1
        else:
            res.append(b[j])
            j += 1
    # Append the remaining tuples
    if i < len(a):
        res += a[i:]
    elif j < len(b):
        res += b[j:]
    return res


def merge_reduce(f, data):
    """ Apply function cumulatively to the items of data,
        from left to right in binary tree structure, so as to
        reduce the data to a single value.
    :param f: function to apply to reduce data.
    :param data: List of items to be reduced.
    :return: result of reduce the data to a single value.
    """
    import queue
    q = queue.Queue()
    for i in data:
        q.put(i)
    while not q.empty():
        x = q.get()
        if not q.empty():
            y = q.get()
            q.put(f(x, y))
        else:
            return x


def main():
    import sys
    import os
    import time
    path = sys.argv[1]

    files_paths = []
    for file in os.listdir(path):
        files_paths.append(path + '/' + file)
    start_time = time.time()
    result = sort_by_key(files_paths)

    print("Elapsed Time(s)")
    print(time.time() - start_time)
    print(result)


if __name__ == "__main__":
    main()
